pagegraph-cli/src/adblock_rules.rs

```diff
@@ -2,9 +2,27 @@
 
 use pagegraph::graph::PageGraph;
 use pagegraph::types::{EdgeType, NodeType};
+use std::fs::File;
+use std::io::{BufReader, BufRead};
 
-pub fn main(graph: &PageGraph, filter_rule: &str) {
-    let matching_elements = graph.resources_matching_filter(&filter_rule);
+pub fn main(graph: &PageGraph, filter_rule: Option<&str>, filter_list_path: Option<&str>) {
+    // Check which one is defined
+    let filter_rules = if let Some(rule) = filter_rule {
+        vec![rule.to_string()]
+    } else {
+        // open file
+        let file = File::open(filter_list_path
+            .expect("At least one of path_to_filterlist or filter_rule must be defined")).unwrap();
```

Comment:
Replacing `filter_rule` and `filter_list_path` with a single enum argument might make this a bit more ergonomic; something like:

```rust
pub enum FilterArg<'a> {
    SingleRule(&'a str),
    PathToList(&'a str),
}
```

Reply:
Hmm but we'd still have to check here which one is defined, right? So we should probably put an Option around each of them and that will lead to this same check? The fact that we're passing in one arg instead of two optional ones is better though, so I'll do that.

Reply:
You won't need an `Option` here at all, I'd add the logic to build the `FilterArg` enum at the CLI level so that this module doesn't have to worry about parsing/validating arguments.

Reply:
I must be missing something - I would still need to check which of `SingleRule` or `PathToList` is defined here right? Given that only one of that will be defined?

Reply:
I decided to sidestep the entire thing by doing the argument check at the CLI level and only passing in the filter list rules (which will only contain one if a filter rule is passed in instead of file) to `adblock_rules`

Reply:
@antonok-edm wdyt?